### PR TITLE
.Net: Pass Kernel to connectors and implement automatic OpenAI function calling

### PR DESCRIPTION
**Pull request by stephentoub**
Closes https://github.com/microsoft/semantic-kernel/issues/3687
Closes https://github.com/microsoft/semantic-kernel/issues/2932

This PR includes three commits:

**Commit 1**: Simply updates XML comments for Kernel arguments to standardize them.

**Commit 2**: Updates the connector interfaces (e.g. IChatCompletion) to accept an optional Kernel argument. It's passed through everywhere, but the actual connector implementations largely ignore it, since they don't need anything from it right now.

**Commit 3**: Updates the {Azure} OpenAI connector to support automatic function calling, whether those functions are explicitly specified in OpenAIPromptExecutionSettings or are implicitly available via the passed in Kernel. The behavior is opt-in, via a revised setting on the OpenAIPromptExecutionSettings class. Rather than just specifying a string (which had been "auto", "none" or a specific function name), this now exposes a property of a new type FunctionCallBehavior, which can be set to various instances to achieve the desired results:
- FunctionCallBehavior.EnableKernelFunctions - the functions in the provided kernel will be sent but won't be automatically invoked if requested
- FunctionCallBehavior.AutoInvokeKernelFunctions - the functions in the provided kernel will be sent and will be automatically invoked if requested
- FunctionCallBehavior.EnableFunctions(functionsList, autoInvoke) - the functions specified will be sent and whether they'll be automatically invoked is determined by the autoInvoke argument and by whether they're available in the provided kernel
- FunctionCallBehavior.RequireFunction(function, autoInvoke) - the specified function will be sent to the service as a request to function call it

The auto-invoke behavior is wired up based on whether autoInvoke is true. The FunctionCallBehavior.AutoInvokeKernelFunctions can also be opted-into in a service-agnostic manner by setting a named property in the base PromptExecutionSettings class's dictionary; future service implementations can respect that as well if they either have a built-in notion of function calling (which some other LLMs do) or if we want to simulate it with a planner baked into the connector implementation.

Please review this both for my API choices and for implementation details. @RogerBarreto, in particular I'd appreciate it if you could look at how streaming is handled. @gitri-ms, I'd appreciate it if you could review the function calling APIs and implementation support.

With all of this, this is now the "Hello, world" of function calling:
```C#
Kernel kernel = new KernelBuilder()
    .WithOpenAIChatCompletion("gpt-3.5-turbo-1106", apiKey)
    .ConfigurePlugins(plugins => plugins.AddPluginFromObject<TimePlugin>())
    .Build();

OpenAIPromptExecutionSettings settings = new() { FunctionCallBehavior = FunctionCallBehavior.AutoInvokeKernelFunctions };
Console.WriteLine(await kernel.InvokePromptAsync("What is the current time?", settings));
```

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.AI.OpenAI/FunctionCallBehavior.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+using System.Collections.Generic;
+using System.Linq;
+using Azure.AI.OpenAI;
+using Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
+
+namespace Microsoft.SemanticKernel.Connectors.AI.OpenAI;
+
+/// <summary>Represents a behavior for OpenAI function calling.</summary>
+public abstract class FunctionCallBehavior
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/FunctionCallBehavior.cs:11]: Function calling has been deprecated in favour of Tools. We are hoping to get an update to the SDK with the Tools API support before v1.0.

Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/FunctionCallBehavior.cs:11]: Thanks. Presumably that would entail changing the existing APIs, anyway, so introducing these now and then revamping later shouldn't make much of a difference?  When do we expect to pull in a new library that has the tools stuff and is there a plan for adapting?

**Review by RogerBarreto — APPROVED** [review]
Thanks for adding this, will benefit on many use cases, 

This is considerably a new architectural change where connectors can trigger function themselfs, making the LLM the actual plan executor.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 2]
```diff
@@ -228,83 +219,230 @@ await foreach (Completions completions in response)
...
+                return responseData.Choices.Select(chatChoice => new ChatResult(responseData, chatChoice)).ToList();
+            }
 
-        if (responseData.Choices.Count == 0)
-        {
-            throw new KernelException("Chat completions not found");
-        }
+            // Get our single result and extract the function call information. If this isn't a function call, or if it is
+            // but we're unable to find the function or extract the relevant information, just return the single result.
+            ChatChoice resultChoice = responseData.Choices[0];
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:262]: That's checked above in the ValidateAutoInvoke at the beginning of the method. If you ask for auto-invocation and multiple responses, it throws an exception.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 3]
```diff
@@ -228,83 +219,230 @@ await foreach (Completions completions in response)
...
+            // Stream any response 
+            Dictionary<string, object>? responseMetadata = null;
+            StringBuilder? contentBuilder = null;
+            string? functionName = null;
+            StringBuilder? functionArgumentsBuilder = null;
+            ChatRole streamedRole = default;
+            CompletionsFinishReason finishReason = default;
+            await foreach (StreamingChatCompletionsUpdate update in response.ConfigureAwait(false))
+            {
+                responseMetadata ??= GetResponseMetadata(update);
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:345]: Yeah, I think this ties in with @markwallace-microsoft's comment that we're going to be bringing in a new SDK with support for tools, and we'll want to switch everything over to that prior to 1.0?

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 4]
```diff
@@ -228,83 +219,230 @@ await foreach (Completions completions in response)
...
 
-        var chatOptions = CreateChatCompletionsOptions(chatRequestSettings, chat, this.DeploymentOrModelName);
+        // Create the Azure SDK ChatCompletionOptions instance from all available information.
+        var chatOptions = CreateChatCompletionsOptions(chatRequestSettings, chat, kernel, this.DeploymentOrModelName);
 
-        Response<ChatCompletions>? response = await RunRequestAsync<Response<ChatCompletions>?>(
-            () => this.Client.GetChatCompletionsAsync(chatOptions, cancellationToken)).ConfigureAwait(false);
-
-        if (response is null)
+        while (true)
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:243]: I'm not following... can you elaborate?

File: dotnet/samples/KernelSyntaxExamples/Example59_OpenAIFunctionCalling.cs [hunk 5]
```diff
@@ -23,203 +22,70 @@ public static class Example59_OpenAIFunctionCalling
...
-        {
-            // Include all functions registered with the kernel.
-            // Alternatively, you can provide your own list of OpenAIFunctions to include.
-            Functions = kernel.Plugins.GetFunctionsMetadata().Select(f => f.ToOpenAIFunction()).ToList(),
-            FunctionCall = "TimePlugin_Date",
-        };
-
-        // Set FunctionCall to the name of a specific function to force the model to use that function.
+        // Set FunctionCall to the result of FunctionCallBehavior.RequireFunction with a specific function to force the model to use that function.
+        executionSettings.FunctionCallBehavior = FunctionCallBehavior.RequireFunction(kernel.Plugins["TimePlugin"]["Date"].Metadata.ToOpenAIFunction(), autoInvoke: true);
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply dotnet/samples/KernelSyntaxExamples/Example59_OpenAIFunctionCalling.cs:47]: I considered that.  The problem, which also exists in this sample but I think should be cleaned up after @markwallace-microsoft lands his change around metadata, is that we'd be passing in an IKernelFunction, which doesn't know its plugin name.  But if we add a `public static KernelFunctionMetadata GetFunctionMetadata(this IReadOnlyPluginCollection plugins, string pluginName, string functionName)` extension method (to go along with the existing `GetFunction` extension method, then we could have RequireFunction overload that takes a KernelFunctionMetadata and do the `ToOpenAIFunction()` if that's desirable.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/FunctionCallBehavior.cs (hunk 1 above)
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/FunctionCallBehavior.cs:11]: The new library version should be ready next week and we do want to include this for v1.0.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:243]: I thought this method implementation got more complex, but I see that there are many return points throughout the logic making `Extract to a Method` more of a pain. All good.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:345]: Not fully sure about that, would be desireable.

**Review by matthewbolanos — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]

**Pull request merged**